README.md

```diff
@@ -1 +1,78 @@
+ vsockHost{"main loop"}
+ eth(("reconstruct ETH frames"))
+ syscall(("OS syscall"))
+ dhcp["DHCP"]
+ dns["DNS"]
+ api["API"]
+portForwarding["Port Forwarding"]
+ vsockHost <----> eth
+ eth <----> syscall
+ vsockHost ----> dhcp
+ vsockHost ----> dns
+ vsockHost ----> api
+ vsockHost ----> portForwarding
+ end
+ end
+ subgraph VM["VM"]
+ subgraph netNs["Isolated Network Namespace"]
+ tapDevice("eth0")
+ subgraph vmSwitch["vm-switch"]
+ vsockVM{"VM Switch"}
+ ethVM(("listens for ETH frames\n from TAP Device"))
+ ethVM <----> vsockVM
+ end
+ tapDevice <----> ethVM
+ init["/sbin/init"]
+ end
+ network-setup["/usr/local/bin/network-setup"]
+ end
+ vsockVM  <---> |AF_VSOCK| vsockHost
+```
+
+## host-switch
+`host-switch` runs on the Windows host and acts as a receiver for all traffic originating from the network namespace within the WSL VM. It performs a handshake to find the right VM to talk to over `AF_VSOCK`. Once a correct VM is found, it then listens for the incoming traffic from that VM. In additon to this, it can provide a DNS resolver that runs in the user space network along with an API that allows for dynamic port forwarding.
+
+## network-setup
+Its main responsibility is to respond to the handshake request from the `host-switch.exe`, create a network namespace and start the `vm-switch` subprocess in the newly created network namespace. Below is a sequence diagram demonstrating the process.
+
+```mermaid
+sequenceDiagram
+    box vm
+    participant script as wsl-init (pid n)
+    participant network-setup
+    participant vm-switch
+    participant init as wsl-init (pid 1)
+    end
+    box host
+    participant host-switch as host-switch.exe
+    end
+
+    script->>network-setup: spawn
+    host-switch->>network-setup: handshake request
+    network-setup->>host-switch: handshake response
+    host-switch->>network-setup: vsock listener ready
+    network-setup->>network-setup: open vsock
+    network-setup->>network-setup: create namespace
+    network-setup->>vm-switch: spawn
+    Note over network-setup,vm-switch: spawn in network namespace, pass in vsock fd
+    network-setup->>init: spawn
+    Note over network-setup,init: spawns in netns, new mnt/pid ns
+    vm-switch->>vm-switch: create lo/eth0
+    vm-switch->>vm-switch: DHCP eth0
+    vm-switch->>vm-switch: listen for connections
+    init-->>init: Spawn /sbin/init
+    vm-switch->>host-switch: forward ethernet
+```
+
+## vm-switch
+Once the `vm-switch` process starts in the new namespace, it creates a tap (`eth0`) and a `lo` device. The Kernel then forwards all the raw Ethernet frames to the tap device. The tap device forwards the frames over [vsock](https://wiki.qemu.org/Features/VirtioVsock) to the host. The process on the host (`host-switch.exe`) decapsulates the frames, and since it maintains both internal (`vm-switch` to `host-switch.exe`) and external (`host-switch.exe` to the internet) connections, it uses syscall to connect to the external endpoints.
```

Comment:
```suggestion
Once the `vm-switch` process starts in the new namespace, it creates a tap (`eth0`) and a `lo` device. The Kernel then forwards all the raw Ethernet frames to the tap device. The tap device forwards the frames over [vsock](https://wiki.qemu.org/Features/VirtioVsock) to the host. The process on the host (`host-switch.exe`) decapsulates the frames, and since it maintains both internal (`vm-switch` to `host-switch.exe`) and external (`host-switch.exe` to the internet) connections, it connects to the external endpoints via normal syscalls.
```